quic/s2n-quic-core/src/recovery/bbr/probe_rtt.rs

```diff
@@ -90,56 +62,87 @@ impl BbrCongestionController {
+        } else if self.timer.is_armed() && round_counter.round_start() {
+            self.round_done = true;
         }
     }
 
-    /// Checks if the ProbeRtt state should be exited and calls `exit_probe_rtt` if so
-    ///
-    /// The next ProbeRtt will be scheduled as necessary
-    pub fn check_probe_rtt_done<Rnd: random::Generator>(
+    /// Returns true if the `ProbeRtt` state is done and should be exited
+    pub fn is_done(&self, now: Timestamp) -> bool {
+        self.round_done && self.timer.is_expired(now)
+    }
+}
+
+/// Methods related to the `ProbeRtt` state
+impl BbrCongestionController {
+    /// Check if it is time to start probing for RTT changes, and enter the ProbeRtt state if so
+    pub fn check_probe_rtt<Rnd: random::Generator>(
         &mut self,
         random_generator: &mut Rnd,
         now: Timestamp,
     ) {
         //= https://tools.ietf.org/id/draft-cardwell-iccrg-bbr-congestion-control-02#4.3.4.4
-        //# BBRCheckProbeRTTDone()
-        //#    if (BBR.probe_rtt_done_stamp != 0 and
-        //#         Now() > BBR.probe_rtt_done_stamp)
-        //#       /* schedule next ProbeRTT: */
-        //#       BBR.probe_rtt_min_stamp = Now()
-        //#       BBRRestoreCwnd()
-        //#       BBRExitProbeRTT()
-
-        if self.probe_rtt_state.timer.poll_expiration(now).is_ready() {
-            /* schedule next ProbeRTT: */
-            self.data_volume_model.schedule_next_probe_rtt(now);
-            self.restore_cwnd();
-            self.exit_probe_rtt(random_generator, now);
+        //# BBRCheckProbeRTT()
+        //#    if (BBR.state != ProbeRTT and
+        //#         BBR.probe_rtt_expired and
+        //#         not BBR.idle_restart)
+        //#       BBREnterProbeRTT()
+        //#       BBRSaveCwnd()
+        //#       BBR.probe_rtt_done_stamp = 0
+        //#       BBR.ack_phase = ACKS_PROBE_STOPPING
+        //#      BBRStartRound()
+        //#     if (BBR.state == ProbeRTT)
+        //#       BBRHandleProbeRTT()
+        //#     if (rs.delivered > 0)
+        //#       BBR.idle_restart = false
+
+        if !self.state.is_probing_rtt()
+            && self.data_volume_model.probe_rtt_expired()
+            && !self.idle_restart
+        {
+            self.state = bbr::State::ProbeRtt(State::default());
+            self.save_cwnd();
+            self.round_counter
+                .set_round_end(self.bw_estimator.delivered_bytes());
+        }
+
+        let probe_rtt_cwnd = self.probe_rtt_cwnd();
+        if let bbr::State::ProbeRtt(probe_rtt_state) = &mut self.state {
+            probe_rtt_state.handle_probe_rtt(
+                &mut self.bw_estimator,
+                &mut self.round_counter,
+                probe_rtt_cwnd,
+                *self.bytes_in_flight,
+                now,
+            );
+            if probe_rtt_state.is_done(now) {
+                self.exit_probe_rtt(random_generator, now);
+            }
```

Comment:
This doesn't match the pseudocode. Is it correct? If so can you add a comment explaining the deviation?

Reply:
In the pseudocode, the probeRTT methods are basically all chained together, like:
```
BBRCheckProbeRTT() -> BBRHandleProbeRTT() -> BBRCheckProbeRTTDone() -> BBRExitProbeRTT()
```

in my impl, I have `BBRHandleProbeRTT()` update the "doneness" of the `ProbeRtt::State`, which can then be checked by calling `probe_rtt_state.is_done()`. I did this in anticipation of an optimization the RFC mentions where when resuming from idle while in the ProbeRTT state, you can check if ProbeRTT is done and exit.  

The RFC has `BBRCheckProbeRTTDone()` updating some additional state, which I've moved into BBRExitProbeRTT(). I can probably add some assertions into `exit_probe_rtt` to double check that the state is done